src/database-controller/src/write-merger/handler.js

```
@@ -0,0 +1,176 @@
+      frameworkName,
+      async () => {
+        const oldFramework = await databaseModel.Framework.findOne({
+          attributes: ['snapshot'],
+          where: { name: frameworkName },
+        });
+        const snapshot = new Snapshot(frameworkRequest);
+        const addOns = new AddOns(
+          configSecretDef,
+          priorityClassDef,
+          dockerSecretDef,
+        );
+        if (!oldFramework) {
+          // create new record in db
+          // including all add-ons and submissionTime
+          // set requestGeneration = 1
+          snapshot.setGeneration(1);
+          const record = _.assign(
+            {},
+            snapshot.getAllUpdate(),
+            addOns.getUpdate(),
+          );
+          record.submissionTime = new Date(submissionTime);
+          await databaseModel.Framework.create(record);
+          return [true, snapshot, addOns];
+        } else {
+          // update record in db
+          const oldSnapshot = new Snapshot(oldFramework.snapshot);
+          // compare framework request (omit requestGeneration)
+          if (
+            _.isEqual(snapshot.getRequest(true), oldSnapshot.getRequest(true))
+          ) {
+            // request is equal, no-op
+            return [false, snapshot, addOns];
+          } else {
+            // request is different
+            // update request in db, mark requestSynced=false
+            snapshot.setGeneration(oldSnapshot.getGeneration() + 1);
+            await databaseModel.Framework.update(
+              _.assign({}, snapshot.getRequestUpdate(), {
+                requestSynced: false,
+              }),
+              { where: { name: frameworkName } },
+            );
+            return [true, snapshot, addOns];
+          }
+        }
+      },
+    );
+    res.status(200).json({ message: 'ok' });
+    // skip db poller, any response or error will be ignored
+    if (needSynchronize) {
+      silentSynchronizeRequest(snapshot, addOns);
```

>silentSynchronizeRequest [](start = 6, length = 24)

Is it async call?
E.g. if silentSynchronizeRequest costs 10min, will the restserver create job call duration contains this 10min?

You can test it

---

Yes. It just creates a promise in backend and we do not await it.

---

Which code line makes the silentSynchronizeRequest to be an async call (or the socalled promise)?

---

It calls `synchronizeRequest`, and `synchronizeRequest` is an async function: https://github.com/microsoft/pai/blob/345d3af77053bbd5d4f0fb61d614715562ee0b20/src/database-controller/src/core/framework.js#L385

An async function returns a promise and will be run in the backend if we don't await it.

---

Ok, so better to rename the silentSynchronizeRequest to be asyncSlientSynchronizeRequest?

---

I don't think it is necessary because async functions are common in Node.js. We usually don't mark a function as `async`, but mark a function as `sync` if it is not an async function. You can take the file system package fs-extra as an reference. https://www.npmjs.com/package/fs-extra#methods

---

So you can comment the async at // skip db poller, any response or error will be ignored ?

---

Commented as belows:

    // Poller has an interval to synchronize the framework request to API server.
    // The interval will cause a delay on every job.
    // So we introduce a short-cut here to synchronize the framework request.
    // It is a async function call and doesn't affect the return of this function.
    // Any error of it will be logged and ignored.

---

ok